Add WebNN backend

pinzhenx opened the pull request.

pinzhenx commented:
@huningxin PTAL, thanks

huningxin commented on lib/session.ts:
```diff
@@ -228,11 +235,29 @@ export class Session {
   }
 
   private initializeOps(graph: Graph): void {
+    graph.partitionBy(this.supportedOps);
```
Do you consider to encapsulate the graph manipulation into `sessionHandler.transformGraph` of WebNN backend?

pinzhenx replied:
I agree. I'll make it more consistent with their original design later.

huningxin commented on lib/session.ts:
```diff
@@ -228,11 +235,29 @@ export class Session {
+          }
+        }
+
+        this._ops[i] = new NNSubgraph(subgraph, initializers, this.enablePseudoReorder, this.prefer, this.profiler);
```
And we can encapsulate above code into `sessionHandler.resolve` of WebNN backend.

huningxin reviewed (commented):
Let's think about a design that introduces `WebNNBackend` and `WebNNSessionHandler`. We can encapsulate the graph transformation into `transformGraph` method and op resolution into `resolve` method of `WebNNSessionHandler`. 

The current onnxjs may not able to partition graph cross backends. So it means WebNN can only run limited models. Let's focus on solo backend execution as first step. BTW, onnxruntime is able to support multiple backends execution. We may discuss and propose similar approach after solo WebNN support is clear. 

What do you think?

pinzhenx commented:
@huningxin This PR is for our experimental use and I completely agree to use the `sessionHandler` if we are going to make it formal.

huningxin commented:
We can keep your branch as our own experiment that serves that purpose very well. Besides that, I also want to explore the more "formal" integration of onnx.js with you. With both, we can discuss with onnxjs owner about the design and plan. What do you think?

pinzhenx commented:
I understand that and I will update the new implementation asap.

pinzhenx commented:
@huningxin It now works in solo mode.

pinzhenx committed:
support cpu/wasm fallback

pinzhenx commented:
@huningxin 

Here's some updates

* onnx.js supports multiple backends
For instance, wasm backend has implemented only a few ops and the rest are falling back to the cpu backend, which is logically using two backends. It's made possible by combining two `RESOLVE_RULES`, i.e., `WASM_OP_RESOLVE_RULES.concat(CPU_OP_RESOLVE_RULES)` and selecting the first match as the impl.

* This PR cannot run with hybrid backends (temporarily)
Though I have added [the arguments](https://github.com/huningxin/onnxjs/pull/1/commits/540cbb497a688e34cb74049751a6a805411f276a#diff-b462f1b901663f16efdf57f208cf732dR23) to enable multiple backends, there were still some issues in the graph, since the current transformed graph was not quite fit their graph execution logic. One way is to hack the `execution-plan` like I've done [before](https://github.com/pinzhenx/onnxjs/blob/webnn_backup/lib/execution-plan.ts). Or I could further transform the graph to work around it. 
Speaking of graph transformation, onnx.js still missing some functions for modifying the graph. So I added some [interfaces](https://github.com/huningxin/onnxjs/pull/1/commits/11bc092d759cc10ad43e831bc704c8758b77ccd2#diff-ee6e5848015d9f99de089a0694d53545R55). Also, it would be better if they support the nested subgraph:

* onnx.js doesn't support subgraph for now
Even if their [doc](https://github.com/onnx/onnx/blob/master/docs/IR.md) defined the nested subgraph , but they [doesn't implement](https://github.com/pinzhenx/onnxjs/blob/webnn_backup/lib/attribute.ts#L122) that yet.

huningxin commented on lib/backends/backend-webnn.ts:
```diff
@@ -0,0 +1,42 @@
+ * WebGLBackend is the entry point for all WebGL opeartions
+ * When it starts it created the WebGLRenderingContext
+ * and other main framework components such as Program and Texture Managers
+ */
```
Please change comments according to WebNN.

huningxin commented on lib/backends/backend-webnn.ts:
```diff
@@ -0,0 +1,42 @@
+  initialize(): boolean {
+    try {
+      this.nnContext = (navigator as NNNavigator).ml.getNeuralNetworkContext();
+      Logger.verbose('WebNNBackend', `Created WebGLContext: ${typeof this.nnContext}`);
```
WebNNContext

huningxin commented on lib/graph.ts:
```diff
@@ -52,6 +52,13 @@ export declare namespace Graph {
+    getOutputIndices(): ReadonlyArray<number>;
+    getOutputNames(): ReadonlyArray<string>;
+    getValues(): ReadonlyArray<Graph.Value>;
+    getNodes(): ReadonlyArray<Graph.Node>;
```
It looks like we only need `setNodes` for webnn backend, does it?

huningxin commented on lib/backends/webnn/nn-subgraph-node.ts:
```diff
@@ -0,0 +1,29 @@
+import {Attribute} from '../../attribute';
+import {Graph} from '../../graph';
+
+export class NNSubgraphNode implements Graph.Node {
```
I prefer to explicit name `WebNNGraphNode`.

huningxin commented on lib/backends/webnn/nn-subgraph-node.ts:
```diff
@@ -0,0 +1,29 @@
+                    .reduce((cnt: any, t: any) => {cnt[t] ? cnt[t]++ : cnt[t]=1; return cnt}, {}))
+                  .map((n: any) => `${n[0]} x ${n[1]}`)
+                  .join(', ') + ` (${this.hashCode()})`;
+    this.opType = 'NNSubgraph';
```
`WebNNGraph`.

huningxin commented on lib/backends/webnn/nn-subgraph-node.ts:
```diff
@@ -0,0 +1,29 @@
+                    .map((node) => node.opType)
+                    .reduce((cnt: any, t: any) => {cnt[t] ? cnt[t]++ : cnt[t]=1; return cnt}, {}))
+                  .map((n: any) => `${n[0]} x ${n[1]}`)
+                  .join(', ') + ` (${this.hashCode()})`;
```
Keep indent aligned to above lines.

huningxin commented on lib/backends/webnn/nn-subgraph-op.ts:
```diff
@@ -0,0 +1,27 @@
+import {NNSubgraphNode} from './nn-subgraph-node';
+import {Graph} from '../../graph';
+
+export class NNSubgraphOp implements Operator {
```
`WebNNGraphOp`

huningxin commented on lib/backends/webnn/op-resolve-rules.ts:
```diff
@@ -0,0 +1,26 @@
+import {NNSubgraphNode} from './nn-subgraph-node';
+
+export const WEBNN_OP_RESOLVE_RULES: ReadonlyArray<OpSet.ResolveRule> = [
+  ['NNSubgraph', '', '1+', (node) => new NNSubgraphOp(node as NNSubgraphNode)],
```
`WebNNGraph`

huningxin commented on lib/backends/webnn/nn-subgraph-op.ts:
```diff
@@ -0,0 +1,27 @@
+    return await inferenceHandler.run(this, inputs);
+  };
+
+  initialize(attributes: Attribute): void {}
```
Could we create and compile WebNN model here?

pinzhenx replied:
It doesn't give any information about the input tensor size during `initialize()`. So I have to compile the graph during the first run,

huningxin commented on lib/backends/webnn/nn-subgraph-op.ts:
```diff
@@ -0,0 +1,27 @@
+
+  initialize(attributes: Attribute): void {}
+
+  checkInputs(inputs: Tensor[]): boolean { return true; }
```
We may need to check inputs. I am fine with a `TODO` comment.

huningxin commented on lib/backends/webnn/nn-subgraph-op.ts:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  async run(inferenceHandler: WebNNInferenceHandler, inputs: Tensor[]): Promise<Tensor[]> {
+    return await inferenceHandler.run(this, inputs);
```
Could we move model execution here?

huningxin commented on lib/backends/webnn/nn-model.ts:
```diff
@@ -0,0 +1,671 @@
+import {NNSubgraphOp} from './nn-subgraph-op';
+import {WebNNInferenceHandler} from './inference-handler';
+
+export class NNModel {
```
`WebNNGraphImpl`

huningxin commented on lib/backends/webnn/nn-model-manager.ts:
```diff
@@ -0,0 +1,44 @@
+ */
+export class NNModelManager {
+
+  repo: Map<NNSubgraphOp, NNModel>;  // this should be per-session object
```
If we can encapsulate `NNModel` into `NNSubgraphOp`, probably we don't need this class.

huningxin commented on lib/backends/webnn/session-handler.ts:
```diff
@@ -0,0 +1,99 @@
+  private opResolveRules: ReadonlyArray<OpSet.ResolveRule>;
+
+  constructor(public readonly backend: WebNNBackend, public readonly context: Session.Context) {
+    const FALLBACK = false;
```
Can we set this true? Or expose this as an option?

huningxin commented on lib/backends/webnn/session-handler.ts:
```diff
@@ -0,0 +1,99 @@
+    });
+    colorGraph.identifyInputOutputTensors(transformer.getInputIndices(), transformer.getOutputIndices());
+    const newNodes = [];
+    for (const {nodeIds, inputIds, outputIds} of colorGraph.partition()) {
```
Can `partition` returns the sub-graph color?

huningxin commented on lib/backends/webnn/session-handler.ts:
```diff
@@ -0,0 +1,99 @@
+      if (v.from === -1 && v.tensor) {
+        const tensor = v.tensor;
+        this.initializers.add(tensorId);
+        this.tensorCache.set(tensorId, TensorUtil.toNHWC(tensor));
```
Can we just keep the reference to `graph` instead of cache all tensors? I concern the memory consumption.

huningxin reviewed (commented):
@pinzhenx , nice job, thanks. I leave some comments for discussion.

pinzhenx committed:
add WebNN Backend

pinzhenx committed:
refactor & support graph partition




x

huningxin commented:
Looks good. Thanks much.

pinzhenx commented on lib/graph.ts:
```diff
@@ -351,7 +376,7 @@ class GraphImpl implements Graph, Graph.Transformer {
           }
           if (data._from !== nodeIndex) {
-            throw new Error(`from property of the Value object doesn't match index of Node being processed`);
+            // throw new Error(`from property of the Value object doesn't match index of Node being processed`);
```
There's a line I forgot to uncomment. It was used to suppress the error by the graph transformation. Now it can be safely uncommented.
